doc/harpoon.txt

```diff
@@ -0,0 +1,281 @@
+6. ⇁ API                                                   *harpoon-⇁-api*
+
+You can define custom behavior of a harpoon list by providing your own calls.
+
+Here is a simple example where i create a list named `cmd` that takes the
+current line in the editor and adds it to harpoon menu. When `list:select(...)`
+is called, we take the contents of the line and execute it as a vim command
+
+I don’t think this is a great use of harpoon, but its meant to show how to
+add your own custom lists. You could imagine that a terminal list would be just
+as easy to create.
+
+>lua
+    local harpoon = require("harpoon")
+    
+    harpoon:setup({
+        -- Setting up custom behavior for a list named "cmd"
+        "cmd" = {
+    
+            -- When you call list:append() this function is called and the return
+            -- value will be put in the list at the end.
+            --
+            -- which means same behavior for prepend except where in the list the
+            -- return value is added
+            --
+            -- @param possible_value string only passed in when you alter the ui manual
+            add = function(possible_value)
+                -- get the current line idx
+                local idx = vim.fn.line(".")
+    
+                -- read the current line
+                local cmd = vim.api.nvim_buf_get_lines(0, idx - 1, idx, false)[1]
+                if cmd == nil then
+                    return nil
+                end
+    
+                return {
+                    value = cmd,
+                    context = { ... any data you want ... },
+                }
+            end,
+    
+            --- This function gets invoked with the options being passed in from
+            --- list:select(index, <...options...>)
+            --- @param list_item {value: any, context: any}
+            --- @param list { ... }
+            --- @param option any
+            select = function(list_item, list, option)
+                -- WOAH, IS THIS HTMX LEVEL XSS ATTACK??
+                vim.cmd(list_item.value)
+            end
+    
+        }
+    })
+<
+
+
+CONFIG                                                *harpoon-⇁-api-config*
+
+There is quite a bit of behavior you can configure via `harpoon:setup()`
+
+- `settings`is the global settings. as of now there isn’t a global setting in use, but once we have some custom behavior i’ll put them here
+- `default`the default configuration for any list. it is simply a file harpoon
+- `[name] = HarpoonPartialConfigItem`any named lists config. it will be merged with `default` and override any behavior
+
+**HarpoonPartialConfigItem Definition**
+
+>
+    ---@class HarpoonPartialConfigItem
+    ---@field encode? (fun(list_item: HarpoonListItem): string)
+    ---@field decode? (fun(obj: string): any)
+    ---@field display? (fun(list_item: HarpoonListItem): string)
+    ---@field select? (fun(list_item?: HarpoonListItem, list: HarpoonList, options: any?): nil)
```

Comment:
I think these should be converted to vimdoc style if possible instead of a code block, probably fine if there's no convenient way to do that though.

Reply:
answered here https://github.com/ThePrimeagen/harpoon/pull/432#discussion_r1430832152